web/lib/supabase/contract-groups.ts

```diff
@@ -0,0 +1,235 @@
+}) {
+  const { userId, isManifoldAdmin, isCreator } = props
+  // if is Manifold Admin or Creator, shows all non private groups in order of popularity
+  if (isManifoldAdmin || isCreator) {
+    const publicGroups = await run(
+      db
+        .from('groups')
+        .select('data')
+        .or('data->>privacyStatus.eq.public,data->>privacyStatus.eq.curated')
+        .order('data->totalMembers', { ascending: false } as any)
+        .limit(GROUPS_LIST_SIZE)
+    )
+
+    return publicGroups.data.map((item) => item.data)
+  } else {
+    // if is not admin/creator, show all non-private groups that they are admin/moderator of
+    const userAdminGroups = await run(
+      db
+        .from('group_role')
+        .select('group_data')
+        .eq('member_id', userId)
+        .or('role.eq.admin,role.eq.moderator')
+        .or(
+          'group_data->>privacyStatus.eq.public,group_data->>privacyStatus.eq.curated'
+        )
+        .order('group_data->totalMembers', { ascending: false } as any)
+        .limit(GROUPS_LIST_SIZE)
+    )
+    return userAdminGroups.data.map((item) => item.group_data)
+  }
+}
+
+export async function searchGroupsToAdd(props: {
+  userId: string
+  isCreator: boolean
+  isManifoldAdmin: boolean
+  prompt: string
+}) {
+  const { userId, isCreator, isManifoldAdmin, prompt } = props
+  if (!prompt || prompt === '') {
+    const data = await getInitialGroupsToAdd({
+      userId: userId,
+      isCreator: isCreator,
+      isManifoldAdmin: isManifoldAdmin,
+    })
+    return data
+  }
+
+  if (isManifoldAdmin) {
+    return await getManifoldAdminGroupsToAdd(prompt)
+  } else if (isCreator) {
+    return await getCreatorGroupsToAdd(prompt, userId)
+  } else {
+    return await getGroupsToAdd(prompt, userId)
+  }
+}
+
+async function getManifoldAdminGroupsToAdd(prompt: string) {
+  //if is manifold admin, select from all groups that are not private
+  const [{ data: exactData }, { data: prefixData }, { data: containsData }] =
+    await Promise.all([
+      run(
+        db
+          .from('groups')
+          .select('*')
+          .or('data->>privacyStatus.eq.public,data->>privacyStatus.eq.curated')
+          .ilike('data->>name', `${prompt}`)
+          .order('data->totalMembers', { ascending: false } as any)
+          .limit(GROUPS_LIST_SIZE)
+      ),
+      run(
+        db
+          .from('groups')
+          .select('*')
+          .or('data->>privacyStatus.eq.public,data->>privacyStatus.eq.curated')
+          .ilike('data->>name', `${prompt}%`)
+          .order('data->totalMembers', { ascending: false } as any)
+          .limit(GROUPS_LIST_SIZE)
+      ),
+      run(
+        db
+          .from('groups')
+          .select('*')
+          .or('data->>privacyStatus.eq.public,data->>privacyStatus.eq.curated')
+          .ilike('data->>name', `%${prompt}%`)
+          .order('data->totalMembers', { ascending: false } as any)
+          .limit(GROUPS_LIST_SIZE)
+      ),
+    ])
```

Comment:
I think a helper function could make this easier.

You could build most of the query once except for the `.ilike()` and then use a helper function to do the parallel queries, e.g.:

```
function (q, field, prompt) {
  return await Promise.all([
  q.ilike(field, prompt),
  q.ilike(field, `{prompt}%`),
  q.ilike(field, `%{prompt}`%),
}
```

Reply:
could not do :(